File: 206-reverse-linked-list/reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
//  rverse the list iteratively and recursively
// 1 -> 2 -> 3 -> 4 -> 5

var reverseList = function(head) {

    // null <- 1 2 -> 3
    //         ^ ^
    let prev = null
    let current = head

    while (current) {
        const tmp = current.next
        current.next = prev
        prev = current
        current = tmp
    }
    return prev
    
};

File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # input: head of a singly linked list
        # output: head of the reversed list
        if head is None:
            return
        if head and head.next is None:
            return head
        
        if head.next:
            new_head = self.reverseList(head.next)
            head.next.next = head
            head.next = None
        
        return new_head

        # return prev

        
            